app/src/main/java/com/hedvig/app/feature/embark/passages/externalinsurer/retrieveprice/RetrievePriceActivity.kt

```diff
@@ -63,9 +61,23 @@ class RetrievePriceInfoActivity : BaseActivity() {
                         )
                     }
                 ) {
+                    val events = viewModel.events.collectAsState(initial = null)
+
+                    when (val event = events.value) {
+                        is RetrievePriceViewModel.Event.Error -> {
+                            ErrorDialog(
+                                onDismiss = {},
+                                message = stringResource(id = event.errorResult.getStringResource())
+                            )
+                        }
+                        is RetrievePriceViewModel.Event.AuthInformation -> {
+                            InsurelyDialog.newInstance(event.reference)
+                                .show(supportFragmentManager, AuthenticateDialog.TAG)
+                        }
+                    }
```

Comment:
Collecting events as State is kinda scary. And more specifically using this state inside the when() statement which is inside the compose scope (not inside a lambda) means that if this recomposes for whatever reason (always assume it will) it will trigger InsurelyDialog.show more than once. You could test this by animating something in the composable for example and see .show get called a gazillion times 😅
So the suggestion here is two-fold:
* Never do side-effects inside the composable scope if we're not in a lambda which is not composable (Inside the onClick: () -> Unit on a Button for example is a perfectly fine place to have side-effects)
* Do not collect events as state and do something like this instead:
```kotlin
LaunchedEffect(Unit) {
    viewModel.events.collect { event -> // or collectLatest if you don't care about potentially dropping older events
        when(event) { 
            // etc 
        }
    }
}
```
I am not 100% sure this approach is flawless or if it can be improved further, but it's probably a good first step until and if we find any reason not to do it this way.

Reply:
And since then you'd be on non-compose land, and you'd still want to show the ErrorDialog, I'd probably flip a boolean like: `val showError by rememberSaveable { mutableStateOf(false) }` (Maybe saveable so that the dialog doesn't close itself on configuration change etc.? I think a simple `remember` would lose state on config change)
And then also flip it again inside the `onDismiss = {}` to make it dismissable by changing the local state.

Reply:
Collecting events in a SideEffect looks correct. The rememberSaveable parts are a bit trickier - when you show a dialog, it will generally survive configuration changes, but not process death.

Reply:
Is that necessarily true though? If we're presenting some important dialog to them, and they get a phone call, talk for X minutes and come back, do we want the dialog to disappear?